File: MemoryTrackerObj/Model/ratingModel/WinnerData.m
//
//  WinnerData.m
//  MemoryTrackerObj
//
//  Created by Alejandro Del Rio Albrechet on 10/4/17.
//  Copyright © 2017 Kristina Del Rio Albrechet. All rights reserved.
//

#import "WinnerData.h"

@implementation WinnerData

- (instancetype) init {
    self = [super init];

    self.username = @"Unknown";
    self.score = 0;
    self.winDate = [[NSDate alloc] init];
    
    return self;
}

- (WinnerData *) initWithUserName: (NSString*) name andScore: (NSInteger) score {
    self = [super init];
    
    self.username = [[NSString alloc] initWithString:name];
    self.score = [[NSNumber alloc] initWithInteger:score];
    self.winDate = [[NSDate alloc] init];
    
    return self;
}

- (WinnerData *)initWithCoder:(NSCoder *)decoder {
    self = [super init];
    
    self.username = [NSString alloc];
    self.score = [NSNumber alloc];
    
    self.username = [decoder decodeObjectForKey:@"username"];
    self.score = [decoder decodeObjectForKey:@"score"];
    self.winDate = [decoder  decodeObjectForKey:@"date"];
    
    return self;
}

- (void)encodeWithCoder:(NSCoder *)encoder {
    [encoder encodeObject:self.username forKey:@"username"];
    [encoder encodeObject:self.score forKey:@"score"];
    [encoder encodeObject:self.winDate forKey:@"date"];
}

- (NSComparisonResult)compare:(WinnerData *)object {
    return !(self.score > object.score);
}

@end

File: MemoryTrackerObj/Model/ratingModel/WinnerData.swift
//
//  WinnerData.swift
//  MemoryTrackerObj
//
//  Created by Alejandro Del Rio Albrechet on 11/1/17.
//  Copyright © 2017 Kristina Del Rio Albrechet. All rights reserved.
//

import Foundation

@objc class WinnerData: NSObject, NSCoding, Comparable {
    
    let username: String
    let score: Int
    let winDate: Date
    
    override init() {
        username = "Unknown"
        score = 0
        winDate = Date()
    }
    
    init(user: String, score: Int) {
        self.username = user
        self.score = score
        winDate = Date()
    }
    
    required init?(coder aDecoder: NSCoder) {
        username = aDecoder.decodeObject(forKey: "username") as? String ?? ""
        score = aDecoder.decodeInteger(forKey: "score")
        winDate = aDecoder.decodeObject(forKey: "date") as? Date ?? Date()
    }
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(username, forKey: "username")
        aCoder.encode(score, forKey: "score")
        aCoder.encode(winDate, forKey: "date")
    }
    
    static func < (lhs: WinnerData, rhs: WinnerData) -> Bool {
        return lhs.score < rhs.score
    }
    
    static func == (lhs: WinnerData, rhs: WinnerData) -> Bool {
        return lhs.score == rhs.score
    }
}


